### PR TITLE
consensus/pbft: re-design message set

### PR DESCRIPTION
**Pull request by alanchchen**
using `*message` and passing validator set

### CODE REVIEW

File: consensus/pbft/core/message_set.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+	for _, v := range ms.messages {
+		result = append(result, v)
+	}
+
+	return result
+}
+
+func (ms *messageSet) Size() int {
+	return len(ms.messages)
```
Comment (bailantaotao) [inline consensus/pbft/core/message_set.go]: Should we add rwMutex RLock here?

File: consensus/pbft/core/hash.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/ethereum/go-ethereum/crypto/sha3"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+func hash(v interface{}) (h common.Hash) {
+	hw := sha3.NewKeccak256()
+	rlp.Encode(hw, v)
+	hw.Sum(h[:0])
+
+	return
```
Comment (markya0616) [inline consensus/pbft/core/hash.go]: remove it

File: consensus/pbft/core/message_set.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		return pbft.ErrNoMatchingValidator
+	}
+
+	// TODO: check view number and sequence number
+
+	return nil
+}
+
+func (ms *messageSet) addVerifiedMessage(msg *message) error {
+	ms.messages[hash(msg)] = msg
```
Comment (markya0616) [inline consensus/pbft/core/message_set.go]:
```ms.messagesMu.Lock()
ms.messagesMu.Unlock()```

File: consensus/pbft/core/message_set.go (hunk 1 above)
Comment (alanchchen) [inline consensus/pbft/core/message_set.go]: Yeah, I'll add it.

File: consensus/pbft/core/message_set.go (hunk 3 above)
Comment (alanchchen) [inline consensus/pbft/core/message_set.go]: Locking is already done in `Add`

File: consensus/pbft/core/hash.go (hunk 2 above)
Comment (alanchchen) [inline consensus/pbft/core/hash.go]:
```
func hash(v interface{}) (h common.Hash)
```
`return` can't be removed, right?

File: consensus/pbft/validator/default.go [hunk 4]
```diff
@@ -51,17 +51,8 @@ type defaultSet struct {
 	proposer   pbft.Validator
 }
 
-func newDefaultSet(extraData []byte) (*defaultSet, bool) {
+func newDefaultSet(addrs []common.Address) (*defaultSet, bool) {
 	valSet := &defaultSet{}
-	if !valSet.CheckFormat(extraData) {
```
Comment (alanchchen) [inline consensus/pbft/validator/default.go]: ExtractValidator

File: consensus/pbft/core/message_set.go [hunk 5]
```diff
@@ -79,8 +92,30 @@ func (ms *messageSet) Size() int {
 // Stream. It is not forbidden to read less or more, but it might
 // be confusing.
 func (ms *messageSet) DecodeRLP(s *rlp.Stream) error {
-	// TODO
-	return nil
+	var sms storageMessageSet
+	err := s.Decode(&sms)
+	if err == nil {
+		valSet, _ := validator.NewSet(sms.Validators)
```
Comment (alanchchen) [inline consensus/pbft/core/message_set.go]: handle error

File: consensus/pbft/core/message_set.go [hunk 6]
```diff
@@ -79,8 +92,30 @@ func (ms *messageSet) Size() int {
...
+
+		if len(sms.Keys) != len(sms.Messages) {
```
Comment (alanchchen) [inline consensus/pbft/core/message_set.go]: move down

**Pull request merged**